# **Discriminative Feature Selection**

# FEATURE SELECTION

Feature Selection is the process where you automatically or manually select those features which contribute most to your prediction variable or output in which you are interested in. Having irrelevant features in your data can decrease the accuracy of the models and make your model learn based on irrelevant features.

We are going to understand it with a practice example. Steps are as follows :

>1) Import important libraries

>2) Importing data

>3) Data Preprocessing

>>i) Price

>>ii) Size

>>iii) Installs

>4) Discriminative Feature Check

>>i) Reviews

>>ii) Price

**1. Import Important Libraries**

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np


In [2]:
#from google.colab import drive
#drive.mount('/content/drive')

**2. Importing Data**

Today we will be working on a playstore apps dataset with ratings. Link to the dataset --> https://www.kaggle.com/lava18/google-play-store-apps/data

In [3]:
df = pd.read_csv('googleplaystore.csv',encoding='unicode_escape')
df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hid...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

**3. Data Preprocessing**

Let us have a look at all the datatypes first :

In [4]:
df.dtypes

App                object
Category           object
Rating            float64
Reviews            object
Size               object
Installs           object
Type               object
Price              object
Content Rating     object
Genres             object
Last Updated       object
Current Ver        object
Android Ver        object
dtype: object

We see that all the columns except 'Rating' are object datatype. We want those columns also as numeric as they dont make sense when they are in object form.Let us start with the 'Price' column.

**i) Price** 

When we saw the head of the dataset, we only see the 0 values in 'Price' column. Let us have a look at the rows with non zero data. As the 'Price column is object type, we compare the column with '0' instead of 0. 

In [5]:
df[df['Price']!='0'].head()

App       Category  Rating  \
234  TurboScan: scan documents and receipts in PDF       BUSINESS     4.7   
235                 Tiny Scanner Pro: PDF Doc Scan       BUSINESS     4.8   
290  TurboScan: scan documents and receipts in PDF       BUSINESS     4.7   
291                 Tiny Scanner Pro: PDF Doc Scan       BUSINESS     4.8   
427                             Puffin Browser Pro  COMMUNICATION     4.0   

    Reviews                Size  Installs  Type  Price Content Rating  \
234   11442                6.8M  100,000+  Paid  $4.99       Everyone   
235   10295                 39M  100,000+  Paid  $4.99       Everyone   
290   11442                6.8M  100,000+  Paid  $4.99       Everyone   
291   10295                 39M  100,000+  Paid  $4.99       Everyone   
427   18247  Varies with device  100,000+  Paid  $3.99       Everyone   

            Genres    Last Updated  Current Ver Android Ver  
234       Business  March 25, 2018        1.5.2  4.0 and up  
235       Business  April 11, 2017        3.4.6  3.0 and up  
290       Business  March 25, 2018        1.5.2  4.0 and up  
291       Business  April 11, 2017        3.4.6  3.0 and up  
427  Communication    July 5, 2018  7.5.3.20547  4.1 and up

We see that the 'Price' column has dollar sign in the beginning for the apps which are not free. Hence we cannot directly convert it to numeric type. We will first have to remove the $ sign so that all datas are uniform and can be converted.

We use the replace function over here to replace the dollar sign by blank. Notice that we had to convert the column to string type from object type as the replace function is only applicable on string functions.

In [6]:
df['Price'] = df['Price'].str.replace('$','')
df[df['Price']!='0'].head()

<ipython-input-6-a2a650a36113>:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will*not* be treated as literal strings when regex=True.
  df['Price'] = df['Price'].str.replace('$','')


App       Category  Rating  \
234  TurboScan: scan documents and receipts in PDF       BUSINESS     4.7   
235                 Tiny Scanner Pro: PDF Doc Scan       BUSINESS     4.8   
290  TurboScan: scan documents and receipts in PDF       BUSINESS     4.7   
291                 Tiny Scanner Pro: PDF Doc Scan       BUSINESS     4.8   
427                             Puffin Browser Pro  COMMUNICATION     4.0   

    Reviews                Size  Installs  Type Price Content Rating  \
234   11442                6.8M  100,000+  Paid  4.99       Everyone   
235   10295                 39M  100,000+  Paid  4.99       Everyone   
290   11442                6.8M  100,000+  Paid  4.99       Everyone   
291   10295                 39M  100,000+  Paid  4.99       Everyone   
427   18247  Varies with device  100,000+  Paid  3.99       Everyone   

            Genres    Last Updated  Current Ver Android Ver  
234       Business  March 25, 2018        1.5.2  4.0 and up  
235       Business  April 11, 2017        3.4.6  3.0 and up  
290       Business  March 25, 2018        1.5.2  4.0 and up  
291       Business  April 11, 2017        3.4.6  3.0 and up  
427  Communication    July 5, 2018  7.5.3.20547  4.1 and up

**ii) Size**

As we see the 'Size' column, we see that the value ends with the letter 'M' for mega. We want to convert the size to numeric value to use in the dataset. Hence we will need to remove the letter 'M'.

For this, we convert the column to string and omit the last letter of the string and save the data in 'Size' column.

Notice from the previous head that we saw, that the 'Size' for row 427 is given as varies with device. We obviously cannot convert such data to numeric. We will see how to deal with it later.

In [7]:
df['Size'] = df['Size'].str[:-1]
df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hid...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews Size     Installs  Type Price Content Rating  \
0     159   19      10,000+  Free     0       Everyone   
1     967   14     500,000+  Free     0       Everyone   
2   87510  8.7   5,000,000+  Free     0       Everyone   
3  215644   25  50,000,000+  Free     0           Teen   
4     967  2.8     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

**iii) Installs**

If we see the 'Installs' column, there are 2 major changes that we need to make to convert it to numeric. We have to remove the '+' sign from the end of the data as well as remove the commas before converting to numeric.

To remove the last letter, we apply the same procedure as for the 'Size' column :

In [8]:
df['Installs'] = df['Installs'].str[:-1]
df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hid...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews Size    Installs  Type Price Content Rating  \
0     159   19      10,000  Free     0       Everyone   
1     967   14     500,000  Free     0       Everyone   
2   87510  8.7   5,000,000  Free     0       Everyone   
3  215644   25  50,000,000  Free     0           Teen   
4     967  2.8     100,000  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

For the removal of commas, we will use the replace function to replace commas with blank.

Replace function only works on string, hence we access the values of the series as string before applying the replace function :

In [9]:
df['Installs'] = df['Installs'].str.replace(',','')
df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hid...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews Size  Installs  Type Price Content Rating  \
0     159   19     10000  Free     0       Everyone   
1     967   14    500000  Free     0       Everyone   
2   87510  8.7   5000000  Free     0       Everyone   
3  215644   25  50000000  Free     0           Teen   
4     967  2.8    100000  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

Now, we will finally convert all the data to numeric type using the to_numeric function. Notice that we have used the errors='coerce' parameter. This parameter converts all the data which cannot be converted to numeric into NaN. For example the 'Size' in row 427 cannot be converted to int. Hence it will be converted to NaN. After that we take a look at the datatypes of the columns again.

In [10]:
df['Reviews'] = pd.to_numeric(df['Reviews'],errors='coerce')
df['Size'] = pd.to_numeric(df['Size'],errors='coerce')
df['Installs'] = pd.to_numeric(df['Installs'],errors='coerce')
df['Price'] = pd.to_numeric(df['Price'],errors='coerce')
df.dtypes

App                object
Category           object
Rating            float64
Reviews           float64
Size              float64
Installs          float64
Type               object
Price             float64
Content Rating     object
Genres             object
Last Updated       object
Current Ver        object
Android Ver        object
dtype: object

Now we will see and work with all the NaN values. Let us first have a look at all the NaN values in the dataset :

In [11]:
df.isna().sum()

App                  0
Category             0
Rating            1474
Reviews              1
Size              1696
Installs             2
Type                 1
Price                1
Content Rating       1
Genres               0
Last Updated         0
Current Ver          8
Android Ver          3
dtype: int64

As rating is the output of our dataset, we cannot have that to be NaN. Hence we will remove all the rows with 'Rating' as NaN :

In [12]:
df = df[df['Rating'].isna()==False]
df.isna().sum()

App                  0
Category             0
Rating               0
Reviews              1
Size              1638
Installs             1
Type                 0
Price                1
Content Rating       1
Genres               0
Last Updated         0
Current Ver          4
Android Ver          3
dtype: int64

This is the final preprocessed dataset that we obtained :

In [13]:
df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hid...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

    Reviews  Size    Installs  Type  Price Content Rating  \
0     159.0  19.0     10000.0  Free    0.0       Everyone   
1     967.0  14.0    500000.0  Free    0.0       Everyone   
2   87510.0   8.7   5000000.0  Free    0.0       Everyone   
3  215644.0  25.0  50000000.0  Free    0.0           Teen   
4     967.0   2.8    100000.0  Free    0.0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

**4. Discriminative Feature Check**

Now we will move on to checking the discriminative feature checking, to see which feature is good and which is not. We will start with the 'Reviews' column. For our case, we will take rating > 4.3 as a good rating. We take that value because as we see in the following stats, the rating is divided 50:50 at that value.

Before we do that, let us have a look at the statistics of the whole table :

In [14]:
df.describe()

Rating       Reviews         Size      Installs        Price
count  9367.000000  9.366000e+03  7729.000000  9.366000e+03  9366.000000
mean      4.193338  5.140498e+05    37.284513  1.789744e+07     0.960928
std       0.537431  3.144042e+06    93.509493  9.123822e+07    15.816585
min       1.000000  1.000000e+00     1.000000  1.000000e+00     0.000000
25%       4.000000  1.862500e+02     6.100000  1.000000e+04     0.000000
50%       4.300000  5.930500e+03    16.000000  5.000000e+05     0.000000
75%       4.500000  8.153275e+04    37.000000  5.000000e+06     0.000000
max      19.000000  7.815831e+07   994.000000  1.000000e+09   400.000000

**i) Reviews**

We will have to check for multiple values that which of them has the best rating distinction. We will start by comparing with the mean of the 'Reviews' column which is 514098.

We will use a new function over here known as crosstab. Crosstab allows us to have a frequency count across 2 columns or conditions.

We could also normalize the column results to obtain the conditional probability of P(Rating = HIGH | condition)

We have also turned on the margins to see the total frequency under that condition.

In [22]:
pd.crosstab(df['Rating']>4.3,df['Reviews']>514098,rownames=['Ratings>4.3'],colnames=['Reviews>514098'],margins= True)

Reviews>514098  False  True   All
Ratings>4.3                      
False            4967   335  5302
True             3383   682  4065
All              8350  1017  9367

We see that the number of ratings in the case of Reviews > 514098 is very less (close to 10%).

Hence it is preferred to take the 50 percentile point rather than the mean to be the pivot point. Let us now take the 50 percentile point which is 5930 reviews in this case. So let us take a look at that :

In [16]:
pd.crosstab(df['Rating']>4.3,df['Reviews']>5930,rownames=['Ratings>4.3'],colnames=['Reviews>5930'],margins= True)

Reviews>5930  False  True   All
Ratings>4.3                    
False          2906  2396  5302
True           1778  2287  4065
All            4684  4683  9367

Now we see that the number of ratings is equal for both high and low reviews. So we will take the 50 percentile point to start from now on. Let us now look at the conditional probability :

In [17]:
pd.crosstab(df['Rating']>4.3,df['Reviews']>5930,rownames=['Ratings>4.3'],colnames=['Reviews>5930'],margins= True,normalize='columns')

Reviews>5930    False      True      All
Ratings>4.3                             
False         0.62041  0.511638  0.56603
True          0.37959  0.488362  0.43397

There is not much difference between P(Ratings=HIGH|Reviews<5930) and P(Ratings=HIGH|Reviews>5930) so this is a bad feature.

Let us increase the value of the pivot for ratings to 80000 and check again. We dont need to check for the percentage being too low as we are almost at 75 percentile mark.

In [18]:
pd.crosstab(df['Rating']>4.3,df['Reviews']>80000,rownames=['Ratings>4.3'],colnames=['Reviews>80000'],margins= True,normalize='columns')

Reviews>80000     False     True      All
Ratings>4.3                              
False          0.613442  0.42518  0.56603
True           0.386558  0.57482  0.43397

Now we see that there is a good difference in the probabilities and hence Rating>80000 is a good feature.

**ii) Price**

We will do the same for 'Price' column to find out the best distinctive feature. We see that in this case, even the 75 percentile mark also points to 0. Hence in this case, we will classify the data as Free or not :

In [19]:
pd.crosstab(df['Rating']>4.3,df['Price']==0,rownames=['Ratings>4.3'],colnames=['Price=$0'],margins= True)

Price=$0     False  True   All
Ratings>4.3                   
False          288  5014  5302
True           360  3705  4065
All            648  8719  9367

This shows us that it is very difficult to use the Price as a feature. Hence it is a doubtful feature. If then also we want to force this as a feature, let us see the conditional probability :

In [20]:
pd.crosstab(df['Rating']>4.3,df['Price']==0,rownames=['Ratings>4.3'],colnames=['Price=$0'],margins= True,normalize='columns')

Price=$0        False      True      All
Ratings>4.3                             
False        0.444444  0.575066  0.56603
True         0.555556  0.424934  0.43397

We see that there is not much difference in probability either, hence this would serve as a bad feature in any case.

This is the end of this tutorial. Now you can move on to assignment 7 in which you have to check the other 2 distinctive features.